executing/executing.py

```diff
@@ -530,6 +539,110 @@ def visit_ClassDef(self, node):
         self.stack.pop()
 
 
+class PositionNodeFinder(object):
+    def __init__(self, frame, stmts, tree, lasti, source):
+        # we can use co_positions() since 3.11, which has fewer limitations
+
+        bc_list = list(dis.Bytecode(frame.f_code, show_caches=True))
+
+        def find_node(index, ignore_positions=(), typ=None):
+            position = bc_list[index // 2].positions
+
+            def match(a, b):
+                for idx, attr in enumerate(
+                    ["lineno", "end_lineno", "col_offset", "end_col_offset"]
+                ):
+
+                    if attr not in ignore_positions and a[idx] != b[idx]:
+                        return False
+
+                return True
+
+
+            return only(
+                node
+                for node in source._nodes_by_line[position[0]]
+                if isinstance(
+                    node, (ast.expr, ast.stmt, ast.excepthandler, ast.MatchAs)
+                )
+                if not isinstance(node, ast.Expr)
+                if typ is None or isinstance(node, typ)
+                if match(
+                    position,
+                    (
+                        node.lineno,
+                        node.end_lineno,
+                        node.col_offset,
+                        node.end_col_offset,
+                    ),
+                )
+            )
+
+        try:
+                node = find_node(lasti)
+        except Exception as e:
+            # lineno of LOAD_METHOD instructions get set to end_lineno by the python compiler
+            # propably to achive improved error messages (PEP-657)
+            # we ignore here the start position and try to find the ast-node just by end position and expected node type
+            if bc_list[lasti // 2].opname in ("LOAD_METHOD","LOAD_ATTR"):
```

Comment:
A few things:

1. I think expanding the comment further would be helpful for future readers.
2. I think STORE_ATTR and DELETE_ATTR are also relevant.
3. Please add some thorough dedicated tests for this stuff testing behaviour with attributes spread across multiple lines. The less we rely on testing sys.modules the better.
4. Can we do better than just ignoring `lineno`? Don't we expect `positions.lineno` to equal `node.end_lineno`?
5. Can we do something similar for `col_offset`? What exactly does CPython set it to?

Reply:
Thank you for your feed back. I have implemented most of it and hope I can implement the tests soon.

I'm not sure about point 4 and 5, here is one example:

``` python
a=(some().
   method(
   with,
   args)
```
I dont know exactly how it is implemented currently, but the current implementation should be save.
I dont know about any case where an other ast-node could end at the same place like the function except binary operations like:
``` python
a().b() + c()
```
end of `c()` is equal to the end of the `+` operation.
and this is also the reason why I limit the search to Call/Attribute.

I will add this explanation to the code. 

Relying on the exact way python changes the position could be quit fragile in the future.

Reply:
> I'm not sure about point 4 and 5, here is one example:

I don't know what you're showing me here. Is this a case where `positions.lineno != node.end_lineno`?

> Relying on the exact way python changes the position could be quit fragile in the future.

Unfortunately this package is fragile by nature and constantly relying on implementation details that might change. I generally prefer it to be as strict as possible, as I consider silently identifying the wrong node to be much worse than refusing to identify a node at all because of excess caution.

Reply:
> I don't know what you're showing me here. Is this a case where positions.lineno != node.end_lineno?
exactly. The `lineno`/`col_offset` is modified during compilation and should be (my not verified opinion) the dot of the method call. This has not to be on the `end_lineno` (like in the example). 

To be as strict as possible is general a good approach, however:

The current implementation can only be broken if there are two `Call` expressions with the same end_lineno/end_col_offset. This is not possible because every Call expression has its own closing brace at the end.
We would have to check for a impossible scenario if we want to be more strict.

The same reasoning can be applied for attribute access. Two `Attribute` expressions can not have the same end because it would mean that there last character would be the same which is just not possible to write.

Please correct me if you see there any mistake in my reasoning.

Reply:
Your reasoning is correct if everything is working properly. But sometimes weird things happen:

- The source code has changed and no longer matches the bytecode
- Other magic such as IPython or pytest interferes. (By the way, your node finder isn't ignoring pytest assertions, we need to check if it still works within them)
- CPython has bugs or just unexpected behaviour that changes over time.

Extra checks assert that the current 'situation' is fully understood and not misleading us.

In https://github.com/python/cpython/blob/330f1d58282517bdf1f19577ab9317fa9810bf95/Python/compile.c#L4719, it looks like `lineno` is simply set to `end_lineno`. I think you can just assume that in general for attributes. `col_offset` only looks slightly messier.

A bit further down at line 4733, `end_col_offset` may get changed to `col_offset`, so relying on it might not be safe.

Reply:
sorry that I did not responded but I got corona. But I was lucky and it was not so bad for me.

Back to the topic:

> it looks like lineno is simply set to end_lineno

no, end_lineno of `attr` not the `node`

``` python
a \
.b \
.c \    # <- this line
(15)
```

and the col_offset gets set to the start of the attribute
``` python
a.method(15)
  ^ should be there
```

> A bit further down at line 4733, end_col_offset may get changed to col_offset, so relying on it might not be safe.

Only if everything is on the same line and col_offset is larger than end_col_offset.
In other words if the end of the source range comes before the start.

> your node finder isn't ignoring pytest assertions
yes you are right we should try to support pytest assertion rewriting.
I think it should be enough to ignore identifiers/nodes which start with @. I think that was used by pytest to prevent name clashes. Just an Idea now, I don't know if it would work perfectly.

I think we would need the asttokens (which is still a optional dependency if I'm correct) to get the required information to recalculate and verify the `lineno` end `col_offset`. I see no other way to get the start of the attribute over the ast.

It is also difficult to test the verification without a real world problem that it should solve.
I currently see no big problem here, also because we could add a verification for a real problem as soon as we find one.

But this is your project and you decide. I will try to build a verification if you want.

Reply:
OK, there's no need to block this PR on this. I can maybe add these extra checks afterwards. Opened https://github.com/alexmojaki/executing/issues/44 for that.

Is there any reason to only specify `matching_nodes` after catching an exception?

Reply:
good question.

I assume you meant the `match_positions` argument.

it is actually only useful for `CALL` instructions because they could be generated from a normal function call, in which case the would be found in the try block. The exception handles then only the case where we call a method. We could also ignore `lineno` and `col_offset` for function calls but I prefer to be a little bit more strict here if possible.

`LOAD_METHOD` and `*_ATTR` could be moved before the try block because we know that we will not find a node if we search with all `matching_nodes`

That are currently just my thoughts. I will try it later.

Reply:
How about something like this:

```python
if opname in ...:
    match_positions = ...

node = find_node(...)

if isinstance(node, ast.Call) and (node.lineno != position.lineno or node.col_offset != position.col_offset):
    assert_(isinstance(node.func, ast.Attribute))
```

Reply:
I tried to reduce it to one `find_node` call but there is a problem.

``` python
        find_args = {}
        if self.opname(lasti) in (
            "LOAD_METHOD",
            "LOAD_ATTR",
            "STORE_ATTR",
            "DELETE_ATTR",
        ):
            find_args = dict(
                match_positions=("end_col_offset", "end_lineno"),
                typ=ast.Attribute,
            )

        try:
            # try to map with all match_positions
            node = self.find_node(lasti, **find_args)
            if self.opname(lasti) == "CALL":
                assert_(isinstance(node,ast.Call),ast.dump(node) )
        except NotOneValueFound:
            # A CALL instruction can be a method call, in which case the lineno and col_offset gets changed by the compiler.
            # Therefore we ignoring here this attributes and searchnig for a Call-node only by end_col_offset and end_lineno.
            # This is save, because there can only be one method ending at a specific point in the source code.
            # One closing ) only belongs to one method.
            if self.opname(lasti) == "CALL":
                node = self.find_node(
                    lasti,
                    match_positions=("end_col_offset", "end_lineno"),
                    typ=ast.Call,
                )
            else:
                raise
```
 
I added a `assert_` to find the reason and there are CALL instructions which map to decorators.
The decorator is a Attribute lookup in this case.

``` python
  raise AssertionError(str(message))
AssertionError: Attribute(value=Name(id='tester', ctx=Load()), attr='qwe', ctx=Load())
```

If we want to move the case for `CALL` up we would have to include anything into `typ` which can be used as a decorator. but then we would find two nodes for cases like `a+b()`

So I think we should keep the second `find_node` call for `CALL` instructions.